test/SlowTests/Issues/RavenDB_16378_Replication_NRE.cs

```diff
@@ -0,0 +1,74 @@
+﻿using System;
+using System.IO;
+using System.Threading.Tasks;
+using FastTests.Server.Replication;
+using Raven.Client.Documents.Operations;
+using Raven.Client.Documents.Operations.Backups;
+using Xunit;
+using Xunit.Abstractions;
+
+namespace SlowTests.Issues
+{
+    public class RavenDB_16378_Replication_NRE : ReplicationTestBase
+    {
+        public RavenDB_16378_Replication_NRE(ITestOutputHelper output) : base(output)
+        {
+        }
+
+        [Fact]
+        public async Task ShouldNotThrowNreOnReplication()
+        {
+            var backupPath = NewDataPath(forceCreateDir: true);
+            var fullBackupPath = Path.Combine(backupPath, "RavenDB-16378-NRE.ravendb-snapshot");
+
+            using (var file = File.Create(fullBackupPath))
+            {
+                // this is a snapshot backup of database which is not completely "valid"
+                // it was created a bit artificially by trying hard to break things when playing with revisions and output reduce results
+                // for example some it includes some artificial documents which don't have @artificial flag (modified manually)
+                // the point of using this snapshot here is that it reproduces the NRE problem (or assertion in AssertNoReferenceToThisPage when running in Debug)
+
+                using (var stream = typeof(RavenDB_16378).Assembly.GetManifestResourceStream("SlowTests.Data.RavenDB_16378.RavenDB-16401-NRE.ravendb-snapshot"))
+                {
+                    stream.CopyTo(file);
+                }
+            }
+
+            using (var mainStore = GetDocumentStore())
+            {
+                var restoredDatabaseName = GetDatabaseName();
+
+                using (RestoreDatabase(mainStore, new RestoreBackupConfiguration
+                {
+                    BackupLocation = backupPath,
+                    DatabaseName = restoredDatabaseName
+                }))
+                using (var sourceOfRestoredDb = GetDocumentStore(new Options
+                {
+                    ModifyDatabaseName = s => restoredDatabaseName,
+                    CreateDatabase = false
+                }))
+                using (var destination = GetDocumentStore(new Options
+                {
+                    ModifyDatabaseName = s => $"{s}_dst"
+                }))
+                {
+                    using (var session = sourceOfRestoredDb.OpenSession())
+                    {
+                        session.Store(new { Foo = "marker"}, "marker");
+
+                        session.SaveChanges();
+                    }
+
+                    await SetupReplicationAsync(sourceOfRestoredDb, destination);
+
+                    WaitForDocument(destination, "marker");
+
+                    var stats = destination.Maintenance.Send(new GetStatisticsOperation());
+
+                    Assert.Equal(6, stats.CountOfDocuments);
```

Comment:
Some explanation - here we're having 6 documents because some artificials has been migrated as well. Although there isn't `orders/2` document because the replication batch contained the following items:

```
(etag: 133) - orders/2 (Document) flags - HasRevisions, DeleteRevision - 3/24/2021 6:57:17 PM
(etag: 137) - orders/2 (Document) flags - HasRevisions, FromReplication - 3/24/2021 6:55:10 PM
(etag: 138) - orders/2 (Document) flags - HasRevisions, Revision, FromReplication - 3/24/2021 6:55:10 PM
```

As the result of resolve to latest conflict resolution strategy the one from `3/24/2021 6:57:17 PM` has won.

Again, the snapshot isn't valid but it reproduced real issue.